rs/src/connections/host_relay.rs

```diff
@@ -0,0 +1,1052 @@
+
+        let config = self.config.clone();
+        tokio::spawn(async move {
+            let session = match russh::server::run_stream(config, rw, server_session).await {
+                Ok(s) => s,
+                Err(e) => {
+                    log::error!("error handshaking session: {}", e);
+                    return Err(e);
+                }
+            };
+
+            if authed_tx.await.is_err() {
+                log::debug!("connection closed before auth");
+                return Ok(()); // session closed
+            }
+
+            let mut known_ports: PortMap = HashMap::new();
+            tokio::pin!(session);
+
+            loop {
+                tokio::select! {
+                    r = &mut session => return r,
+                    cnx = server_connection_rx.recv() => match cnx {
+                        Some(cnx) => {
+                            if let Some(p) = known_ports.get(&cnx.port) {
+                                p.send(cnx).ok(); // ignore error, could have dropped in the meantime
+                            }
+                        },
+                        None => {
+                            log::debug!("no more connections, ending");
+                            return Ok(());
+                        },
+                    },
+                    _ = ports.changed() => {
+                        let new_ports = ports.borrow().clone();
+                        for port in new_ports.keys() {
+                            if !known_ports.contains_key(port) {
+                                session.handle().forward_tcpip("127.0.0.1".to_string(), *port).await.ok();
+                            }
+                        }
+                        for port in known_ports.keys() {
+                            if !new_ports.contains_key(port) {
+                                session.handle().cancel_forward_tcpip("127.0.0.1".to_string(), *port).await.ok();
+                            }
+                        }
+
+                        known_ports = new_ports;
+                    },
+                }
+            }
+        })
+    }
+}
+
+/// Connects connections that are sent to the receiver to TCP services locally.
+/// Runs until the receiver is closed (usually via `delete_port()`).
+async fn forward_port_to_tcp(
+    addr: impl tokio::net::ToSocketAddrs + std::fmt::Display,
+    mut rx: mpsc::UnboundedReceiver<ForwardedPortConnection>,
+) {
+    while let Some(mut conn) = rx.recv().await {
+        let mut stream = match TcpStream::connect(&addr).await {
+            Ok(s) => s,
+            Err(e) => {
+                log::info!("Error connecting forwarding to {}, {}", addr, e);
+                conn.close().await;
+                continue;
+            }
+        };
+
+        log::debug!("Forwarded port to {}", addr);
+
+        tokio::spawn(async move {
+            let mut read_buf = [0u8; 1024 * 64];
```

Comment:
I'm a bit concerned about allocating 64k of stack. (There's been some discussion of how we want to tune the unstable `large_assignment` lint [here](https://github.com/rust-lang/rust/issues/83518#issuecomment-1072779300) and 64k is called out as "huge")

I would suggest using `vec![0u8; 1024 * 64].into_boxed_slice()` to get a `Box<[u8]>` or something like [this](https://stackoverflow.com/a/72504064/124538) if you want a `Box<[u8; 1024 * 64]>`.